docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,341 @@
+        if grid.north_edge:
+            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
+        if grid.east_edge:
+            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)
+
+This example motivates a number of requirements for the feature. It must be able to
+
+1. specify regions in the context of a global iteration space (for global boundary conditions) using different `with` blocks inside a single stencil
+2. specify arbitrary sub-regions using a clean abstraction
+3. specify cleanly regions at and near the edge of compute domains
+4. read from and write to fields inside halo regions
+
+The specific feature that we are proposing is the addition of a `region()` specifier that complements the `interval()`, but treats the horizontal instead of the vertical.
+Arguments passed to the `region()` specifier can be arbitrary Python expressions resolved at compile time, which result in objects specifying the local index region for the contained computation.
+In order to make this friendlier for applications, they can provide methods or classes that provide these objects specifying the local index region.
+
+Using this concept, the example above is transformed into:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type) # Single stencil
+    def ubke(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
+        with computation(PARALLEL):
+            with interval(...): # applies to the compute domain as it does now
+                ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with region((WestEdge(0, 1), EastEdge(0, 1)), interval(...):
+                # Executes on regions defined by the Box objects
+                ub = dt5 * (ut[0, -1, 0] + ut)​
+            with region(SouthEdge(0, 1), NorthEdge(0, 1)), interval(...):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+    # Python wrapper to unpack arguments
+    def compute(uc, vc, ut, ub, dt5, dt4, grid):​
+        ubke(uc, vc, grid.cosa, grid.rsina, ub, ut, dt5, dt4)
+
+This reduces the complexity of the code and consolidates operations on `ub` - it is now clear what the stencil is doing to `ub` everywhere.
+Model developers already understand the idea of offsets with the stencil notation of e.g. `uc[0, -1, 0]`, which is analogous to specifying offsets from an edge of the horizontal domain.
+
+All that needs to be changed in the frontend is adding the `region()` specifier, **so all existing GT4Py code will continue to work as before.**
+The new `region()` specifier accepts `local_origin` and `extent` that describe the horizontal iteration space in a way that describes the process-local iteration space, and can also represent the global space if the space is domain decomposed and computed by a set of processes.
```

Comment:
and if you get fields with different halo sizes the question is if the box is relative to the (global) origin or relative to the end of the compute domain of any given field?

Reply:
This is an issue I've struggled with, as well. The issue is that there doesn't really exist a "east edge", there's only a mid-level east edge and an interface-level east edge index. We can handle this either by using EastEdge and knowing to add offsets for interface-level variables, or by using `EastEdge` and `InterfaceEastEdge` or somesuch. There are still inevitable issues when you use both mid and interface variables with the latter, though.

This is in some ways a limitation underlying gt4py - that you can define an index iteration space and it is to be applied uniformly across a set of variables (the inputs to a stencil), even though those variables may have differing shapes/sizes.

I think in short this is a larger issue in gt4py, the GDP suits the current way gt4py operates and making it possible for stencil offsets to be array-shape or dimensionally aware is a separate issue.

Reply:
> how do you intend to deal with fields that have different sizes?

Good question. It would be done the same way as it is now. Application developers set the domain and origin for a field, and loop over some indices in gt4py. So `LeftEdge(0,1)` corresponds to some global indices, converted to local, then again interpreted potentially differently when indexing the field and accessing the underlying memory.

Reply:
> and if you get fields with different halo sizes the question is if the box is relative to the (global) origin or relative to the end of the compute domain of any given field?

0 should always remain relative to the global origin. If you want to use negative indices, that's at the application's own risk.